lib/internal/streams/async_iterator.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+const { promisify } = require('util');
+
+class Item {
+  constructor(value, done) {
+    this.done = done;
+    this.value = value;
+  }
+}
+
+function readAndResolve(iter) {
+  if (iter.lastResolve !== null) {
+    const data = iter.stream.read();
+    // we defer if data is null
+    // we can be expecting either 'end' or
+    // 'error'
+    if (data !== null) {
+      iter.lastResolve(new Item(data, false));
+      iter.lastResolve = null;
+      iter.lastReject = null;
+    }
+  }
+}
+
+class ReadableAsyncIterator {
+  constructor(stream) {
+    this.stream = stream;
+    this.lastResolve = null;
+    this.lastReject = null;
+    this.error = null;
+    this.ended = false;
+
+    stream.on('readable', () => {
+      // we wait for the next tick, because it might
+      // emit an error with process.nextTick
+      process.nextTick(readAndResolve, this);
+    });
+
+    stream.on('end', () => {
```

Comment:
I think, in general, `bind` would be better here for performance.

That said, this makes me wish we already had `WeakRefs` in JS because we could just make `stream[kAsyncIterator]` and get rid of closures altogether.